<a href="https://colab.research.google.com/github/tamamori3180-hash/GCI-/blob/main/%E3%82%BF%E3%82%A4%E3%82%BF%E3%83%8B%E3%83%83%E3%82%AF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 目標
　Titanic号の乗客に関する様々な情報から、**その乗客が最終的に亡くなったかどうか**をできるだけ正しく予測してください。乗客の死亡/生存情報は、Perishedカラムの値に対応付けられています。Perishedカラムの値が1であれば死亡した、0であれば生存したということに対応しています。本コンペはオリジナルの"Titanic: Machine Learning from Disaster"コンペを模してはいるものの同一ではないため、一部カラムの命名や意味が異なります。下記のデータの項をよく読んで取り組むようにしてください。

## データ
　データ分析コンペティションでは、一般的に**訓練データ**と**テストデータ**が与えられます。今回の場合、訓練データは**data/train.csv**、テストデータは**data/test.csv**となっています。

訓練データとテストデータは以下のような点で異なります。  
- 訓練データ：このデータによって機械学習モデルを学習させます。
- テストデータ：このデータに対する予測を提出して、その精度を競います。   

　含まれているデータの種類は以下のようになっています。  

変数 |定義 |備考  
---|---|---
Perished |死亡したかどうか |0 = No, 1 = Yes
Pclass |チケットのクラス |1 = 1st, 2 = 2nd, 3 = 3rd
Name |名前 |
Sex |性別 |
Age	|年齢 |
SibSp |乗船していた兄弟姉妹・配偶者の数	|
Parch |乗船していた親・子供の数	|
Ticket |チケット番号	|
Fare |チケット料金	|
Cabin |キャビン番号	|
embarked |乗船した港	|C = Cherbourg, Q = Queenstown, S = Southampton

#環境構築

In [ ]:
# Google driveをマウント
# Google ColabでGoogleドライブを使えるようにするためのもの
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# 参照ディレクトリの設定
# Google Colab上で「作業するフォルダ（ディレクトリ）」を指定して、そこに移動するためのもの
%cd "/content/drive/MyDrive"

/content/drive/MyDrive


In [ ]:
# データ処理用
import pandas as pd  # データフレーム操作

# 可視化用
import matplotlib.pyplot as plt  # 可視化
import seaborn as sns  # 可視化

In [ ]:
# 外部ライブラリのインストール
!pip install japanize-matplotlib # 日本語対応（Matplotlibで日本語を表示）

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.1/4.1 MB 48.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for japanize-matplotlib: filename=japanize_matplotlib-1.1.3-py3-none-any.whl size=4120257 sha256=aed65fdf81e5e26c1464750b42f38fa626c46525673024290ce33bfa4b1663c4
  Stored in directory: /root/.cache/pip/wheels/c1/f7/9b/418f19a7b9340fc16e071e89efc379aca68d40238b258df53d
Successfully built japanize-matplotlib


In [ ]:
# 外部ライブラリのインポート
import japanize_matplotlib  # 日本語対応（Matplotlibで日本語を表示）

In [ ]:
# 訓練データ読み込み
df = pd.read_csv("GCI/winter/competition_1/data/train.csv", encoding = "latin-1")
dft = pd.read_csv("GCI/winter/competition_1/data/test.csv", encoding = "latin-1")

#訓練データ

##欠損値

In [ ]:
df

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,0,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,0,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,0,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,0,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,1,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,0,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Perished     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df.isnull().sum()

,0
Perished,0
Pclass,0
Sex,0
Age,177
SibSp,0
Parch,0
Fare,0
Cabin,687
Embarked,2


Age, Cabin, Embarkedにおいて欠損値が確認できた。

###Age

In [ ]:
df[df["Age"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
5,6,1,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
17,18,0,2,"Williams, Mr. Charles Eugene",male,NaN,0,0,244373,13.0000,NaN,S
19,20,0,3,"Masselmani, Mrs. Fatima",female,NaN,0,0,2649,7.2250,NaN,C
26,27,1,3,"Emir, Mr. Farred Chehab",male,NaN,0,0,2631,7.2250,NaN,C
28,29,0,3,"O'Dwyer, Miss. Ellen ""Nellie""",female,NaN,0,0,330959,7.8792,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
859,860,1,3,"Razi, Mr. Raihed",male,NaN,0,0,2629,7.2292,NaN,C
863,864,1,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,NaN,8,2,CA. 2343,69.5500,NaN,S
868,869,1,3,"van Melkebeke, Mr. Philemon",male,NaN,0,0,345777,9.5000,NaN,S
878,879,1,3,"Laleff, Mr. Kristo",male,NaN,0,0,349217,7.8958,NaN,S


Nameに含まれる'Miss'と'Mrs'という敬称で平均年齢にズレがあるかも

In [ ]:
def extract_title(name):
    """名前から敬称（Mr, Miss, Mrsなど）を抽出する"""
    title_search = name.split(', ')[1].split('.')[0].strip()
    return title_search

# 1. Nameカラムから敬称（タイトル）を抽出し、新しいカラム'Title'を作成
df['Title'] = df['Name'].apply(extract_title)

# 2. 'Miss'と'Mrs'のみをフィルタリングし、平均年齢を計算
average_ages = df[df['Title'].isin(['Miss', 'Mrs'])].groupby('Title')['Age'].mean()

# 3. 結果を表示
print(average_ages)

Title
Miss    21.773973
Mrs     35.898148
Name: Age, dtype: float64


In [ ]:
df

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,Mr
1,2,0,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,Mrs
2,3,0,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,Miss
3,4,0,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S,Mrs
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S,Mr
...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S,Rev
887,888,0,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S,Miss
888,889,1,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S,Miss
889,890,0,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C,Mr


In [ ]:
dfn = df.drop(columns=["Name", "Ticket", "PassengerId"])
dfn

,Perished,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin,Embarked,Title
0,1,3,male,22.0,1,0,7.2500,NaN,S,Mr
1,0,1,female,38.0,1,0,71.2833,C85,C,Mrs
2,0,3,female,26.0,0,0,7.9250,NaN,S,Miss
3,0,1,female,35.0,1,0,53.1000,C123,S,Mrs
4,1,3,male,35.0,0,0,8.0500,NaN,S,Mr
...,...,...,...,...,...,...,...,...,...,...
886,1,2,male,27.0,0,0,13.0000,NaN,S,Rev
887,0,1,female,19.0,0,0,30.0000,B42,S,Miss
888,1,3,female,NaN,1,2,23.4500,NaN,S,Miss
889,0,1,male,26.0,0,0,30.0000,C148,C,Mr


In [ ]:
dfn.groupby('Title')['Age'].median()

,Age
Title,
Capt,70.0
Col,58.0
Don,40.0
Dr,46.5
Jonkheer,38.0
Lady,48.0
Major,48.5
Master,3.5
Miss,21.0


敬称としての「Master」
「Mr.」の由来: 過去には、社会的な一人前の男性に「Mr.」が使われる以前は、「master」が敬称として使われていました。
「Mstr.」: 現在も「Mr.」の略称として「Mstr.」が使われることがあります。
少年に対する敬称: 現代では、成人男性には「Mr.」が使われ、少年（子供）に対しては「Master」が敬称として残っています。

In [ ]:
# 3. 'Title' ごとに中央値を算出し、その値で Age の NaN を補完
# groupby().transform('median') が最も効率的で推奨される方法です。

# グループごとの中央値を計算し、それを Age の NaN 部分に代入
dfn['Age'].fillna(dfn.groupby('Title')['Age'].transform('median'), inplace=True)

# 補完後の Age の欠損値の数を確認（0になっているはずです）
print("Ageカラムの欠損値補完後の残りの数:", dfn['Age'].isna().sum())

Ageカラムの欠損値補完後の残りの数: 0


/tmp/ipython-input-285645711.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  dfn['Age'].fillna(dfn.groupby('Title')['Age'].transform('median'), inplace=True)


In [ ]:
dfn['Age'].isna()

,Age
0,False
1,False
2,False
3,False
4,False
...,...
886,False
887,False
888,False
889,False


In [ ]:
dfn1 = dfn.drop(columns=["Title"])
dfn1

,Perished,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin,Embarked
0,1,3,male,22.0,1,0,7.2500,NaN,S
1,0,1,female,38.0,1,0,71.2833,C85,C
2,0,3,female,26.0,0,0,7.9250,NaN,S
3,0,1,female,35.0,1,0,53.1000,C123,S
4,1,3,male,35.0,0,0,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...
886,1,2,male,27.0,0,0,13.0000,NaN,S
887,0,1,female,19.0,0,0,30.0000,B42,S
888,1,3,female,21.0,1,2,23.4500,NaN,S
889,0,1,male,26.0,0,0,30.0000,C148,C


In [ ]:
dfn1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Perished  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       891 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


###Cabin

In [ ]:
df[df["Cabin"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,0,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,1,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
7,8,1,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
884,885,1,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,1,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
888,889,1,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S


In [ ]:
# Cabin も NaN であることが分かる。データを確認すると空欄だった。
# Age NaN 行を削除する

df_drop1 = df.dropna(subset=['Age'])
df_drop1[df_drop1["Age"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked


In [ ]:
# Age NaN を削除した Cabin の欠損値

df_drop1[df_drop1["Cabin"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,0,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
7,8,1,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,0,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
883,884,1,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.5000,NaN,S
884,885,1,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.0500,NaN,S
885,886,1,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S


In [ ]:
# Pclass=3 において欠損していると仮定した

df_drop1_s = df_drop1[(df_drop1["Pclass"] == 3)]

Pclass = 3 である行を抜き出したが、Cabin の欠損値の半分ほどであったため関連性はない？

In [ ]:
# Perished=1 死が Cabin に影響していると仮定

df_drop1_s1 = df_drop1[(df_drop1["Perished"] == 1)]

In [ ]:
df_drop1_s1[df_drop1_s1["Cabin"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.250,NaN,S
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.050,NaN,S
7,8,1,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.075,NaN,S
12,13,1,3,"Saundercock, Mr. William Henry",male,20.0,0,0,A/5. 2151,8.050,NaN,S
13,14,1,3,"Andersson, Mr. Anders Johan",male,39.0,1,5,347082,31.275,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
883,884,1,2,"Banfield, Mr. Frederick James",male,28.0,0,0,C.A./SOTON 34068,10.500,NaN,S
884,885,1,3,"Sutehall, Mr. Henry Jr",male,25.0,0,0,SOTON/OQ 392076,7.050,NaN,S
885,886,1,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.125,NaN,Q
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.000,NaN,S


Perished=1 である行を抜き出したが、Cabin の欠損値の半分ほどであったため関連性はない？

Cabin の欠損値がデータ全体891行の内529行と約60%を占めている為、削除することは望ましくないと考える。

船は、部屋番号により船内のどこに位置しているかが分かる。（一般的に）

よって、カテゴリー別にまとめる。



In [ ]:
df_drop1_encoded = df['Has_Cabin'] = df['Cabin'].apply(lambda x: 1 if pd.notna(x) else 0)
df_drop1_encoded

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,...,Cabin_E68,Cabin_E77,Cabin_E8,Cabin_F G63,Cabin_F G73,Cabin_F2,Cabin_F33,Cabin_F4,Cabin_G6,Cabin_T
0,1,1,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,...,0,0,0,0,0,0,0,0,0,0
1,2,0,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,...,0,0,0,0,0,0,0,0,0,0
2,3,0,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,...,0,0,0,0,0,0,0,0,0,0
3,4,0,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,...,0,0,0,0,0,0,0,0,0,0
4,5,1,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
885,886,1,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,...,0,0,0,0,0,0,0,0,0,0
886,887,1,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,...,0,0,0,0,0,0,0,0,0,0
887,888,0,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,...,0,0,0,0,0,0,0,0,0,0
889,890,0,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
# 最頻値で補完する対象の列
cols_to_fill = ["Cabin", "Embarked"]

# train の最頻値で train/test 両方を補完
for col in cols_to_fill:
    mode_value = dfn1[col].mode()[0]
    dfn1[col] = dfn1[col].fillna(mode_value)

In [ ]:
dfn1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Perished  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       891 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     891 non-null    object 
 8   Embarked  891 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


###Embarked

In [ ]:
df[df["Embarked"].isna()]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,0,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,0,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


Ticket が同一の番号である。

In [ ]:
# Ticket=113572 が他にもないか確認

df[(df["Ticket"] == "113572")]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,0,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,0,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


In [ ]:
# Cabin=B28 が他にもないか確認

df[df["Cabin"] == "B28"]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,0,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,0,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


In [ ]:
# Fare=80 が他にもないか確認

df[df["Fare"] == 80]

,PassengerId,Perished,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,0,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,0,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


欠損値が2行と非常に少ないため、データ全体に対する影響が無視できる程度と考え削除する

##異常値

In [ ]:
df.describe()

,PassengerId,Perished,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.616162,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,1.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


##カテゴリカルデータ

In [ ]:
df.describe(include="obj

,Name,Sex,Ticket,Cabin,Embarked
count,891,891,891,204,889
unique,891,2,681,147,3
top,"Dooley, Mr. Patrick",male,347082,G6,S
freq,1,577,7,4,644


In [ ]:
df.value_counts(["Embarked"])

,count
Embarked,
S,644
C,168
Q,77


#テストデータ

In [ ]:
dft

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...
413,1305,3,"Spector, Mr. Woolf",male,NaN,0,0,A.5. 3236,8.0500,NaN,S
414,1306,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C
415,1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S
416,1308,3,"Ware, Mr. Frederick",male,NaN,0,0,359309,8.0500,NaN,S


In [ ]:
dft.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.1+ KB


Age, Fare, Cabin, Embarkedにおいて欠損値が確認できた。

##欠損値

###Age

In [ ]:
dft = pd.read_csv("GCI/冬/data/test.csv", encoding = "latin-1")
# 提出用に元のPassengerIdを保存
test_passenger_ids = dft['PassengerId'].copy()

In [ ]:
dft[dft["Age"].isna()]

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
10,902,3,"Ilieff, Mr. Ylio",male,NaN,0,0,349220,7.8958,NaN,S
22,914,1,"Flegenheim, Mrs. Alfred (Antoinette)",female,NaN,0,0,PC 17598,31.6833,NaN,S
29,921,3,"Samaan, Mr. Elias",male,NaN,2,0,2662,21.6792,NaN,C
33,925,3,"Johnston, Mrs. Andrew G (Elizabeth Lily"" Watson)""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
36,928,3,"Roth, Miss. Sarah A",female,NaN,0,0,342712,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...
408,1300,3,"Riordan, Miss. Johanna Hannah""""",female,NaN,0,0,334915,7.7208,NaN,Q
410,1302,3,"Naughton, Miss. Hannah",female,NaN,0,0,365237,7.7500,NaN,Q
413,1305,3,"Spector, Mr. Woolf",male,NaN,0,0,A.5. 3236,8.0500,NaN,S
416,1308,3,"Ware, Mr. Frederick",male,NaN,0,0,359309,8.0500,NaN,S


In [ ]:
dft[dft["Age"].isna()]

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title
88,980,3,"O'Donoghue, Ms. Bridget",female,NaN,0,0,364856,7.75,NaN,Q,Ms


In [ ]:
dft

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q,Mr
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S,Mrs
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q,Mr
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S,Mr
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S,Mrs
...,...,...,...,...,...,...,...,...,...,...,...,...
413,1305,3,"Spector, Mr. Woolf",male,28.5,0,0,A.5. 3236,8.0500,NaN,S,Mr
414,1306,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C,Dona
415,1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S,Mr
416,1308,3,"Ware, Mr. Frederick",male,28.5,0,0,359309,8.0500,NaN,S,Mr


In [ ]:
dft.groupby('Title')['Age'].median()

,Age
Title,
Col,50.0
Dona,39.0
Dr,53.0
Master,7.0
Miss,22.0
Mr,28.5
Mrs,36.5
Rev,35.5


In [ ]:
# 欠損値は全て補完されたため、この行は不要です。
# dft = dft.dropna(subset=['Age'])

In [ ]:
dft

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q,Mr
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S,Mrs
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q,Mr
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S,Mr
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S,Mrs
...,...,...,...,...,...,...,...,...,...,...,...,...
413,1305,3,"Spector, Mr. Woolf",male,28.5,0,0,A.5. 3236,8.0500,NaN,S,Mr
414,1306,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C,Dona
415,1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S,Mr
416,1308,3,"Ware, Mr. Frederick",male,28.5,0,0,359309,8.0500,NaN,S,Mr


###Fare

In [ ]:
dft[dft["Fare"].isna()]

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title
152,1044,3,"Storey, Mr. Thomas",male,60.5,0,0,3701,NaN,NaN,S,Mr


In [ ]:
# 中央値で補完する対象の列
cols_to_fill = ["Fare"]

# train の値で train/test 両方を補完
for col in cols_to_fill:
    median_value = dft[col].median()
    dft[col] = dft[col].fillna(median_value)

/tmp/ipython-input-282262076.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dft[col] = dft[col].fillna(median_value)


In [ ]:
dft[dft["Fare"].isna()]

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title


###Cabin

In [ ]:
dft[dft["Cabin"].isna()]

,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...
412,1304,3,"Henriksson, Miss. Jenny Lovisa",female,28.0,0,0,347086,7.7750,NaN,S
413,1305,3,"Spector, Mr. Woolf",male,NaN,0,0,A.5. 3236,8.0500,NaN,S
415,1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,NaN,S
416,1308,3,"Ware, Mr. Frederick",male,NaN,0,0,359309,8.0500,NaN,S


In [ ]:
dft.describe()

,PassengerId,Pclass,Age,SibSp,Parch,Fare
count,418.000000,418.000000,332.000000,418.000000,418.000000,417.000000
mean,1100.500000,2.265550,30.272590,0.447368,0.392344,35.627188
std,120.810458,0.841838,14.181209,0.896760,0.981429,55.907576
min,892.000000,1.000000,0.170000,0.000000,0.000000,0.000000
25%,996.250000,1.000000,21.000000,0.000000,0.000000,7.895800
50%,1100.500000,3.000000,27.000000,0.000000,0.000000,14.454200
75%,1204.750000,3.000000,39.000000,1.000000,0.000000,31.500000
max,1309.000000,3.000000,76.000000,8.000000,9.000000,512.329200


In [ ]:
# 最頻値で補完する対象の列
cols_to_fill = ["Cabin"]

# train の最頻値で train/test 両方を補完
for col in cols_to_fill:
    mode_value = dft[col].mode()[0]
    dft[col] = dft[col].fillna(mode_value)

/tmp/ipython-input-2945069033.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dft[col] = dft[col].fillna(mode_value)


In [ ]:
dft.info()

<class 'pandas.core.frame.DataFrame'>
Index: 417 entries, 0 to 417
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  417 non-null    int64  
 1   Pclass       417 non-null    int64  
 2   Name         417 non-null    object 
 3   Sex          417 non-null    object 
 4   Age          417 non-null    float64
 5   SibSp        417 non-null    int64  
 6   Parch        417 non-null    int64  
 7   Ticket       417 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        417 non-null    object 
 10  Embarked     417 non-null    object 
 11  Title        417 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 42.4+ KB


SibSpの値とParchの値の和をとって「同乗していた家族の人数」という新しい変数を加える

In [ ]:
# 1. 'SibSp'と'Parch'のカラムの値を足し合わせ、新しいカラム 'Family_Size' を作成
# 重要な点: この合計には乗客本人（+1）が含まれていません。
dfn1['Family_Size'] = dfn1['SibSp'] + dfn1['Parch']

# 2. 乗客本人を含めた「同乗していた家族の人数（本人含む）」に修正
# 本人（1）を足して、家族の総人数を算出します。
dfn1['Family_Size'] = dfn1['Family_Size'] + 1

In [ ]:
dfn1

,Perished,Pclass,Sex,Age,SibSp,Parch,Fare,Cabin,Embarked,Family_Size
0,1,3,male,22.0,1,0,7.2500,B96 B98,S,2
1,0,1,female,38.0,1,0,71.2833,C85,C,2
2,0,3,female,26.0,0,0,7.9250,B96 B98,S,1
3,0,1,female,35.0,1,0,53.1000,C123,S,2
4,1,3,male,35.0,0,0,8.0500,B96 B98,S,1
...,...,...,...,...,...,...,...,...,...,...
886,1,2,male,27.0,0,0,13.0000,B96 B98,S,1
887,0,1,female,19.0,0,0,30.0000,B42,S,1
888,1,3,female,21.0,1,2,23.4500,B96 B98,S,4
889,0,1,male,26.0,0,0,30.0000,C148,C,1


In [ ]:
dfn2 = dfn1.drop(columns=["SibSp", "Parch"])
#カテゴリデータを数値に変換
dfn3 = pd.get_dummies(dfn2, columns=["Sex", "Cabin", "Embarked"], dtype=int, drop_first=True)
dfn3

,Perished,Pclass,Age,Fare,Family_Size,Sex_male,Cabin_A14,Cabin_A16,Cabin_A19,Cabin_A20,...,Cabin_F G63,Cabin_F G73,Cabin_F2,Cabin_F33,Cabin_F38,Cabin_F4,Cabin_G6,Cabin_T,Embarked_Q,Embarked_S
0,1,3,22.0,7.2500,2,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
1,0,1,38.0,71.2833,2,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,0,3,26.0,7.9250,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
3,0,1,35.0,53.1000,2,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
4,1,3,35.0,8.0500,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,1,2,27.0,13.0000,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
887,0,1,19.0,30.0000,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
888,1,3,21.0,23.4500,4,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
889,0,1,26.0,30.0000,1,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
dfn3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Columns: 154 entries, Perished to Embarked_S
dtypes: float64(2), int64(152)
memory usage: 1.0 MB


In [ ]:
# 1. 'SibSp'と'Parch'のカラムの値を足し合わせ、新しいカラム 'Family_Size' を作成
# 重要な点: この合計には乗客本人（+1）が含まれていません。
dft['Family_Size'] = dft['SibSp'] + dft['Parch']

# 2. 乗客本人を含めた「同乗していた家族の人数（本人含む）」に修正
# 本人（1）を足して、家族の総人数を算出します。
dft['Family_Size'] = dft['Family_Size'] + 1
dft

/tmp/ipython-input-766105584.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dft['Family_Size'] = dft['SibSp'] + dft['Parch']
/tmp/ipython-input-766105584.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dft['Family_Size'] = dft['Family_Size'] + 1


,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Title,Family_Size
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,B57 B59 B63 B66,Q,Mr,1
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,B57 B59 B63 B66,S,Mrs,2
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,B57 B59 B63 B66,Q,Mr,1
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,B57 B59 B63 B66,S,Mr,1
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,B57 B59 B63 B66,S,Mrs,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
413,1305,3,"Spector, Mr. Woolf",male,28.5,0,0,A.5. 3236,8.0500,B57 B59 B63 B66,S,Mr,1
414,1306,1,"Oliva y Ocana, Dona. Fermina",female,39.0,0,0,PC 17758,108.9000,C105,C,Dona,1
415,1307,3,"Saether, Mr. Simon Sivertsen",male,38.5,0,0,SOTON/O.Q. 3101262,7.2500,B57 B59 B63 B66,S,Mr,1
416,1308,3,"Ware, Mr. Frederick",male,28.5,0,0,359309,8.0500,B57 B59 B63 B66,S,Mr,1


In [ ]:
# 訓練データの前処理と一致するように、特徴量として使用しないカラムを削除します。
# 訓練データでは 'SibSp', 'Parch', 'Name', 'Title', 'PassengerId', 'Ticket' が最終的に削除または処理対象外でした。

cols_to_drop_from_dft = ["SibSp", "Parch", "Name", "Title", "PassengerId"]
# 'Ticket' カラムがまだ存在する場合のみドロップリストに追加（念のため）
if "Ticket" in dft.columns:
    cols_to_drop_from_dft.append("Ticket")

dft_features = dft.drop(columns=cols_to_drop_from_dft)

# カテゴリデータを数値に変換します。訓練データと一致するように 'Sex', 'Cabin', 'Embarked' のみを使用します。
dft_processed = pd.get_dummies(dft_features, columns=["Sex", "Cabin", "Embarked"], dtype=int, drop_first=True)

# 訓練データ (X) のカラムとテストデータ (dft_processed) のカラムを揃えます。
# Xは既にセルmRzCc3Bi1j5Cで定義されているはずです。
train_cols = X.columns

# テストデータに存在しない訓練データのカラムを追加し、0で埋めます。
missing_cols_in_test = set(train_cols) - set(dft_processed.columns)
for col in missing_cols_in_test:
    dft_processed[col] = 0

# テストデータにのみ存在するカラムがあれば削除します。
extra_cols_in_test = set(dft_processed.columns) - set(train_cols)
if len(extra_cols_in_test) > 0:
    dft_processed = dft_processed.drop(columns=list(extra_cols_in_test))

# カラムの順序を訓練データと一致させます。
dft2 = dft_processed[train_cols]

# カラムの一致を確認します
print("Columns in X (training features):", X.columns.tolist())
print("Columns in dft2 (test features):", dft2.columns.tolist())
assert X.columns.tolist() == dft2.columns.tolist(), "Test and train columns do not match!"
print("dft2の最終的な行数と列数:", dft2.shape)
print(dft2.info())

Columns in X (training features): ['Pclass', 'Age', 'Fare', 'Family_Size', 'Sex_male', 'Cabin_A14', 'Cabin_A16', 'Cabin_A19', 'Cabin_A20', 'Cabin_A23', 'Cabin_A24', 'Cabin_A26', 'Cabin_A31', 'Cabin_A32', 'Cabin_A34', 'Cabin_A36', 'Cabin_A5', 'Cabin_A6', 'Cabin_A7', 'Cabin_B101', 'Cabin_B102', 'Cabin_B18', 'Cabin_B19', 'Cabin_B20', 'Cabin_B22', 'Cabin_B28', 'Cabin_B3', 'Cabin_B30', 'Cabin_B35', 'Cabin_B37', 'Cabin_B38', 'Cabin_B39', 'Cabin_B4', 'Cabin_B41', 'Cabin_B42', 'Cabin_B49', 'Cabin_B5', 'Cabin_B50', 'Cabin_B51 B53 B55', 'Cabin_B57 B59 B63 B66', 'Cabin_B58 B60', 'Cabin_B69', 'Cabin_B71', 'Cabin_B73', 'Cabin_B77', 'Cabin_B78', 'Cabin_B79', 'Cabin_B80', 'Cabin_B82 B84', 'Cabin_B86', 'Cabin_B94', 'Cabin_B96 B98', 'Cabin_C101', 'Cabin_C103', 'Cabin_C104', 'Cabin_C106', 'Cabin_C110', 'Cabin_C111', 'Cabin_C118', 'Cabin_C123', 'Cabin_C124', 'Cabin_C125', 'Cabin_C126', 'Cabin_C128', 'Cabin_C148', 'Cabin_C2', 'Cabin_C22 C26', 'Cabin_C23 C25 C27', 'Cabin_C30', 'Cabin_C32', 'Cabin_C45', 'Ca

/tmp/ipython-input-470140624.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dft_processed[col] = 0
/tmp/ipython-input-470140624.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dft_processed[col] = 0
/tmp/ipython-input-470140624.py:21: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dft_proce

In [ ]:
dft2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 417 entries, 0 to 417
Columns: 443 entries, Pclass to Ticket_W.E.P. 5734
dtypes: float64(2), int64(441)
memory usage: 1.4 MB


#データ分割

In [ ]:
# データフレームが df であると仮定します。
# Perished カラムが目的変数（ターゲット）であると想定します。

# 1. 目的変数 (y) の作成
# 'Perished' カラムを抽出します
df_n = dfn3['Perished']

# 2. 特徴量 (X) の作成
# 目的変数 'Perished' カラムをデータフレームから削除（drop）します
# axis=1 は「カラム（列）」を削除することを示します
df_N = dfn3.drop('Perished', axis=1)

In [ ]:
import numpy as np                                   # 数値計算や配列操作
import pandas as pd                                  # 表形式のデータを扱う
import matplotlib.pyplot as plt                      # 基本的なグラフ描画
import seaborn as sns                                # きれいで便利な統計グラフ
from sklearn.preprocessing import LabelEncoder       # カテゴリ変数を数値に変換
from sklearn.ensemble import RandomForestClassifier  # ランダムフォレスト分類器
from sklearn.model_selection import train_test_split # データの分割
from sklearn.metrics import accuracy_score

In [ ]:
# 特徴量と目的変数に分ける
X = dfn3.drop(columns=["Perished"])  # 予測に使う説明変数
y = dfn3["Perished"]                 # 予測したい目的変数

# データを7:3に分割（訓練データ70%、検証データ30%）
X_train, X_valid, y_train, y_valid = train_test_split(
    X, y, test_size=0.3, random_state=46, stratify=y
)

# モデルの設定（ランダムフォレスト）
model = RandomForestClassifier(
    n_estimators=100,
    max_depth=5,
    random_state=2025
)

# モデルを訓練データで学習
model.fit(X_train, y_train)

# 検証データで予測（クラスラベルを出力）
y_valid_pred = model.predict(X_valid)

# Accuracyスコアで性能を評価
accuracy = accuracy_score(y_valid, y_valid_pred)
print("検証データの正解率:", round(accuracy, 4))


検証データの正解率: 0.7985


In [ ]:
dft2

,Pclass,Age,Fare,Family_Size,Sex_male,Cabin_A18,Cabin_A21,Cabin_A29,Cabin_A34,Cabin_A9,...,Ticket_SOTON/OQ 392083,Ticket_STON/O 2. 3101268,Ticket_STON/O 2. 3101291,Ticket_STON/O2. 3101270,Ticket_STON/OQ. 369943,Ticket_W./C. 14260,Ticket_W./C. 14266,Ticket_W./C. 6607,Ticket_W./C. 6608,Ticket_W.E.P. 5734
0,3,34.5,7.8292,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,3,47.0,7.0000,2,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,2,62.0,9.6875,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,3,27.0,8.6625,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,3,22.0,12.2875,3,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
413,3,28.5,8.0500,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
414,1,39.0,108.9000,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
415,3,38.5,7.2500,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
416,3,28.5,8.0500,1,1,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
# 検証データで予測（クラスラベルを出力）
# dft2 のカラムが訓練データと一致するように修正されたため、予測が可能です。
pred = model.predict(dft2)

In [ ]:
# Google Drive・Google Colaboratoryで作業する場合
submission.to_csv('submission.csv',index=False)

from google.colab import files
files.download('submission.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [ ]:
 # 読み込むデータが格納されたディレクトリのパス，必要に応じて変更の必要あり
path =  '/content/drive/MyDrive/GCI/冬/data/'

submission = pd.read_csv(path + 'gender_submission.csv')
submission

,PassengerId,Perished
0,892,1
1,893,0
2,894,1
3,895,1
4,896,0
...,...,...
413,1305,1
414,1306,0
415,1307,1
416,1308,1


In [ ]:
pred

array([1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0,
       1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1,
       1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,

In [ ]:
# 学習済みモデルを使って、testデータに対する予測を行う
test_pred = model.predict(dft2)

# 提出用のCSVファイルを作成
# test_passenger_idsはセルxQ9qGyyVweYKで保存されています。
# dft2の行数とtest_passenger_idsの行数が一致しているはずです。
submission = pd.DataFrame({'PassengerId': test_passenger_ids, 'Perished': test_pred})

# 提出ファイルを保存
# 'path'変数はセルibWRC22x75aZで定義されています。
submission.to_csv(path + 'submission.csv', index=False)

print("提出ファイルが正常に作成されました！")
print(submission.head())

提出ファイルが正常に作成されました！
   PassengerId  Perished
0          892         1
1          893         1
2          894         1
3          895         1
4          896         0
